internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,204 @@
+	for {
+		// Check if Stop has been called
+		select {
+		case <-f.stopCh:
+			f.flushBuffer()
+			return
+		default:
+			// continue
+		}
+
+		if f.buffer.ReadyToFlush() {
+			f.flushBuffer()
+		}
+
+		if f.parser.Next() {
+			f.buffer.WriteEntry(f.parser.Entry())
+		} else {
+			if err := f.parser.Err(); err != nil {
+				logger.Noticef(
+					"Forwarding logs to target %q: cannot parse log for service %q: %v",
+					f.targetName, f.serviceName, err)
+			}
+		}
+	}
+}
+
+func (f *lokiForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *lokiForwarder) flushBuffer() {
+	err := f.client.Send(f.buffer.Request())
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot parse log for service %q: %v",
+			f.targetName, f.serviceName, err)
+		return
+	}
+
+	f.buffer.Empty()
+}
+
+type lokiBuffer struct {
+	labels      map[string]string
+	messages    []string
+	maxMessages int
+
+	deadline *time.Time
+	delay    time.Duration
+}
+
+func newLokiBuffer(labels map[string]string, maxMessages int, delay time.Duration) *lokiBuffer {
+	return &lokiBuffer{
+		labels:      labels,
+		messages:    make([]string, 0, maxMessages),
+		maxMessages: maxMessages,
+		deadline:    nil,
+		delay:       delay,
+	}
+}
+
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    [ "<unix epoch in nanoseconds>", "<log line>" ]
+	timestamp := fmt.Sprint(entry.Time.UnixNano())
+	message := strings.Trim(entry.Message, "\n")
+	encodedMsg := fmt.Sprintf("[%q,%q]", timestamp, message)
+	b.messages = append(b.messages, encodedMsg)
+
+	if b.deadline == nil {
+		// Set write deadline
+		deadline := time.Now().Add(b.delay)
+		b.deadline = &deadline
+	}
+}
+
+func (b *lokiBuffer) ReadyToFlush() bool {
+	return (len(b.messages) >= b.maxMessages) ||
+		(b.deadline != nil && time.Now().After(*b.deadline))
+}
+
+// Prepares a JSON request for sending to Loki.
+func (b *lokiBuffer) Request() io.Reader {
+	labelStrings := []string{}
+	for k, v := range b.labels {
+		labelStrings = append(labelStrings, fmt.Sprintf("%q:%q", k, v))
+	}
+
+	req := &bytes.Buffer{}
+	req.WriteString(fmt.Sprintf(`{"streams":[{"stream":{%s},"values":[%s]}]}`,
+		strings.Join(labelStrings, ","), strings.Join(b.messages, ",")))
+	return req
+}
+
+// Reset the buffer.
+func (b *lokiBuffer) Empty() {
+	b.messages = make([]string, 0, b.maxMessages)
+	b.deadline = nil
+}
+
+const (
+	lokiPushAPI = "/loki/api/v1/push"
+)
+
+// lokiClient handles sending requests to Loki.
+type lokiClient struct {
+	client *http.Client
+	url    string
+}
+
+func newLokiClient(location string) *lokiClient {
+	return &lokiClient{
+		client: &http.Client{},
+		url:    location,
+	}
+}
+
+func (c *lokiClient) Send(body io.Reader) error {
+	resp, err := c.client.Post(c.url+lokiPushAPI, "application/json", body)
+	if err != nil {
+		// TODO: annotate this error with some more context?
+		return err
+	}
+	// Check if the response has error code
+	if resp.StatusCode >= 400 {
```

Comment:
seems like if we get anything other than 200 it should be an error right? otherwise we should handle every "success" code.

Reply:
204 is the usual response from Loki. only 4xx and 5xx reflect an error